### PR TITLE
Adds rhost url support behind a feature flag

### PR DESCRIPTION
**Pull request by dwelch-r7**
This PR is for an additional option that allows users to specify a URL (i.e. `set RHOST_URL https://example.com/path`) rather than setting each individual option that is the current workflow (i.e. `set RHOSTS example.com`, `set RPORT 443`, `set SSL true` ... etc)

For full details on the various approaches to adding URL Support to Metasploit:
https://github.com/rapid7/metasploit-framework/wiki/RFC---Metasploit-URL-support

I currently only have the new option added to `Exploit::Remote::HttpClient` but I'm sure there are a few other places it also belongs, more than happy to add it in anywhere that makes sense

Here is an example of the setting in action:
image

The reverse is also true, i.e. you can set multiple individual values and the RHOST_URL will be displayed for you:
image

Http username and password is also supported:
image

Some things to note with the current implementation:
- Only single URLs may be specified (In other words you are not able specify an address range in the new option), attempting this will blank out the RHOST_URL option, but any previous options that were set will remain.
image

- URLs of the format `example.com` notably missing the scheme are not supported, you must do `//example.com:<port>` or even better `https://example.com`
image

To sum up, if this feature is to be added users would be able to simply input `set RHOST_URL` and paste in their target rather than having to set anywhere up to 6 different options depending on what that particular module cared about.

Continued from #13766.

### CODE REVIEW

File: spec/lib/msf/core/opt_http_rhost_url_spec.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  describe '#calculate_value missing scheme' do
+    calculate_values_no_schema.each do | url |
+
+      context url[:normalized].to_s do
+        it "should return #{url[:value]}" do
+          expect(subject.calculate_value(url[:normalized])).to eq(URI('http://' + url[:value]))
+        end
+      end
+    end
+  end
```
Comment (adfoster-r7) [inline spec/lib/msf/core/opt_http_rhost_url_spec.rb:59]:
```suggestion
  describe '#calculate_value' do
    [
      {
         datastore: { ... },
         expected_url: URI("http://....")
      }
    ].each do |test|
        it "should return #{test[:expected_url]}" do
          expect(subject.calculate_value(test[:datastore])).to eq(test[:expected_uri])
        end
    end
  end
end
```

File: lib/msf/core/opt_raw.rb [hunk 2]
```diff
@@ -28,7 +28,7 @@ def normalize(value)
     value
   end
 
-  def valid?(value=self.value)
+  def valid?(value=self.value, check_empty: true)
```
Comment (adfoster-r7) [inline lib/msf/core/opt_raw.rb:31]: Just confirming if this is needed / is there any context on this change? 👀

**Review by adfoster-r7 — DISMISSED** [review]
This currently crashes for me when I try to use it, full details here https://github.com/rapid7/metasploit-framework/pull/13961#discussion_r469316639

File: lib/msf/core/opt_http_rhost_url.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      # Blank this out since we don't know if this new value will have a `VHOST` to ensure we remove the old value
+      option_hash['VHOST'] = nil
+
+      option_hash['RHOSTS'] = uri.hostname
+      option_hash['RPORT'] = uri.port
+      option_hash['SSL'] = %w[ssl https].include?(uri.scheme)
+
+      # Both `TARGETURI` and `URI` are used as datastore options to denote the path on a uri
+      option_hash['TARGETURI'] = uri.path || '/'
+      option_hash['URI'] = uri.path || '/'
```
Comment (adfoster-r7) [inline lib/msf/core/opt_http_rhost_url.rb:30]: `uri.path` seems to return an empty string, which is truthy in ruby - so I think this defaulting logic will never occur 👀

Comment (dwelch-r7) [reply lib/msf/core/opt_http_rhost_url.rb:30]: yup, will change

File: spec/lib/msf/core/opt_http_rhost_url_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
+require 'msf/core/opt_http_rhost_url'
```
[the remark this answers is not in the document]
Comment (dwelch-r7) [reply spec/lib/msf/core/opt_http_rhost_url_spec.rb:1]: what does this do...?

File: lib/msf/core/opt_http_rhost_url.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      value = 'http://' + value unless value.start_with?(%r{https?://})
+      URI(value)
+    rescue URI::InvalidURIError
+      nil
+    end
+
+    def check_for_range(value)
+      return false if value =~ /[^-0-9,.*\/]/
+      walker = Rex::Socket::RangeWalker.new(value)
+      if walker&.valid?
```
[the remark this answers is not in the document]
Comment (dwelch-r7) [reply lib/msf/core/opt_http_rhost_url.rb:81]: I'm not sure if it can be, but I figured better to be safe? I don't remember if passing it something invalid causes it to be nil or if I was just being cautious at the time

[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/msf/core/opt_http_rhost_url.rb:81]: It looks like it's just a normal constructor method to me, I don't think there's any need for the nil check here in that case :eyes:

File: lib/msf/core/opt_raw.rb (hunk 2 above)
Comment (adfoster-r7) [reply lib/msf/core/opt_raw.rb:31]: Caught up with Dean, this was just making the API consistent with the other opt implementations of `valid?`

File: lib/msf/core/option_container.rb [hunk 6]
```diff
@@ -18,6 +18,7 @@ module Msf
   autoload :OptRaw, 'msf/core/opt_raw'
   autoload :OptRegexp, 'msf/core/opt_regexp'
   autoload :OptString, 'msf/core/opt_string'
+  autoload :OptHTTPRhostURL, 'msf/core/opt_http_rhost_url'
```
[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/msf/core/option_container.rb:21]: Let's leave this for now

File: spec/lib/msf/core/opt_http_rhost_url_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (adfoster-r7) [reply spec/lib/msf/core/opt_http_rhost_url_spec.rb:1]: It's just to be explicit, it's not currently needed with the `rspec` test setup that we have:

https://github.com/rapid7/metasploit-framework/blob/482e146a6eb5ce138644734c7ec568ee85522eff/.rspec#L3

It seems like the majority of our spec files use explicit imports, ~250 with, versus ~50 without.

File: lib/msf/core/opt_http_rhost_url.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      end
+    end
+
+    protected
+
+    def get_uri(value)
+      return unless value
+      return if check_for_range(value)
+
+      value = 'http://' + value unless value.start_with?(%r{https?://})
```
[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/msf/core/opt_http_rhost_url.rb:72]: The rhost url code no longer crashes for me; But there's an unrelated error:

```
msf6 auxiliary(scanner/http/title) > features set RHOST_HTTP_URL true
RHOST_HTTP_URL => true
msf6 auxiliary(scanner/http/title) > use scanner/http/title
reload
set RHOST_HTTP_URL https://github.com/rapid7/metasploit-framework/pull/13961/files
run
msf6 auxiliary(scanner/http/title) > reload
[*] Reloading module...
msf6 auxiliary(scanner/http/title) > set RHOST_HTTP_URL https://github.com/rapid7/metasploit-framework/pull/13961/files
RHOST_HTTP_URL => https://github.com/rapid7/metasploit-framework/pull/13961/files
msf6 auxiliary(scanner/http/title) > run

[+] [140.82.118.4:443] [C:200] [R:] [S:GitHub.com] Adds rhost url support behind a feature flag by dwelch-r7 · Pull Request #13961 · rapid7/metasploit-framework · GitHub
[-] Auxiliary failed: RuntimeError Problem with HTTP request for type/path: POST /api/v1/notes message: "\xC2" from ASCII-8BIT to UTF-8
[-] Call stack:
[-]   /Users/adfoster/Documents/code/metasploit-framework/lib/metasploit/framework/data_service/remote/http/response_data_helper.rb:63:in `json_to_mdm_object'
[-]   /Users/adfoster/Documents/code/metasploit-framework/lib/metasploit/framework/data_service/remote/http/remote_note_data_service.rb:15:in `report_note'
[-]   /Users/adfoster/Documents/code/metasploit-framework/lib/metasploit/framework/data_service/proxy/note_data_proxy.rb:40:in `block in report_note'
[-]   /Users/adfoster/Documents/code/metasploit-framework/lib/metasploit/framework/data_service/proxy/core.rb:166:in `data_service_operation'
[-]   /Users/adfoster/Documents/code/metasploit-framework/lib/metasploit/framework/data_service/proxy/note_data_proxy.rb:38:in `report_note'
[-]   /Users/adfoster/Documents/code/metasploit-framework/lib/msf/core/auxiliary/report.rb:177:in `report_note'
[-]   /Users/adfoster/Documents/code/metasploit-framework/modules/auxiliary/scanner/http/title.rb:91:in `run_host'
[-]   /Users/adfoster/Documents/code/metasploit-framework/lib/msf/core/auxiliary/scanner.rb:117:in `block (2 levels) in run'
[-]   /Users/adfoster/Documents/code/metasploit-framework/lib/msf/core/thread_manager.rb:106:in `block in spawn'
[*] Auxiliary module execution completed
```

File: lib/msf/core/opt_http_rhost_url.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+      nil
+    end
+
+    def single_rhost?(value)
+      return true if value =~ /[^-0-9,.*\/]/
+      walker = Rex::Socket::RangeWalker.new(value)
+      return false unless walker.valid?
+        # if there is only a single ip then it's not a range
+      walker.length == 1
+    rescue ::Exception
```
Comment (adfoster-r7) [inline lib/msf/core/opt_http_rhost_url.rb:84]: Looks like we don't want to capture Exception here? 👀 

https://thoughtbot.com/blog/rescue-standarderror-not-exception

File: spec/lib/msf/core/opt_http_rhost_url_spec.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+      context test[:datastore].to_s do
+        it "should return #{test[:expected_url]}" do
+          expect(subject.calculate_value(test[:datastore])).to eq(test[:expected_url])
+        end
+      end
+    end
+  end
+
+end
```
Comment (adfoster-r7) [inline spec/lib/msf/core/opt_http_rhost_url_spec.rb:41]: Let's rubocop this file

**Comment by adfoster-r7** [comment]
## Release notes

Updates http modules to support the setting of multiple http options a single option value `RHOST_HTTP_URL`. For example:  `set RHOST_HTTP_URL http://example.com/`. This functionality is currently opt-in. The steps for enabling this feature can be seen within the `help features` command.
